### PR TITLE
[None][feat] Add TriAttention KV-cache compression method

### PR DESCRIPTION
**Pull request by Hudayday**
## Summary

- Add TriAttention [NVIDIA work @ ICML 2026](https://icml.cc/virtual/2026/poster/66741) KV-cache compression support to the TensorRT-LLM.
- TriAttention is a training-free decode-time eviction algorithm: every `beta` confirmed generation tokens, it scores the evictable decode region using offline per-head query calibration, keeps `budget` important tokens, compacts their K/V data in place, and releases unused tail pages.
- Integrate TriAttention as a KV-cache compression resource manager on top of `KVCacheManagerV2`; decode continues to use the model's standard attention backend.
- Add the score/select pipeline: V2 page-table snapshot → mean-phase gather → CuTe paged-K scoring → normalize/reduce → TopK/tie settlement → batched physical compaction.
- Support `union`, `per_head`, and `per_layer_perhead` eviction modes.
- Support GPT-OSS-style kernel-masked SWA and one-model MTP/EAGLE3 draft-cache co-compaction in `union` mode.
- Preserve logical RoPE positions after physical eviction by disabling RoPE fusion for eviction-based compression.
- Add configuration/API wiring, usage documentation, B200 CI registration, and focused tests.

## Performance

On Qwen3-8B, TriAttention (B=2048, β=2048) delivers up to 1.61× the throughput of the Dense baseline at the same batch size (BS=32: 4,930.6 vs. 3,057.5 output tok/s per GPU; 154.1 vs. 95.5 tok/s per user), while its verified capacity frontier reaches 2.63× the Dense baseline’s maximum aggregate throughput (8,053.0 vs. 3,057.5 output tok/s per GPU).
Screenshot 2026-07-29 at 10 56 06 AM

## Main Files

| File | Purpose |
|---|---|
| `_torch/kv_cache_compression/triattention/triattention.py` | Request lifecycle, calibration loading, eviction cadence, score/select/compact orchestration, and cache resize |
| `triattention_cute_score_fused.py` | SM100 CuTe DSL paged-K score kernel and score pipeline |
| `triattention_cute_selection.py` | CuTe DSL score normalization and union reduction |
| `triattention_kernels.py` | Triton mean-phase, per-head reduction, TP-fold, and tie-settlement kernels |
| `batch_manager/kvCacheManagerV2Utils.{h,cpp}` | Batched V2 page-table snapshot used by the scorer and compactor |
| `llmapi/llm_args.py` | `TriAttentionKvCacheCompressionConfig` and KV-compression config dispatch |
| `examples/triattention/README.md` | Calibration, configuration, support matrix, and usage |

## Key Design Decisions

- **Physical compression, not sparse attention:** TriAttention changes which K/V entries are stored. It does not introduce a sparse-attention backend or modify the normal decode attention kernel.
- **In-place V2 compaction:** The implementation does not create a new KV-cache instance or memory pool. Selected K/V entries are moved within the existing paged pools through the generic compaction API from #16836, then the same cache is resized.
- **Direct paged-pool scoring:** The scorer consumes a batched snapshot of the current V2 page table and reads paged K directly; no Torch gather/reformat path is used.
- **Position correctness:** Prompt tokens, selected-token RoPE information, and protected overlap/speculative tails are preserved. `py_num_compressed_tokens` maintains the logical sequence length after the physical cache becomes shorter.
- **Reusable runtime:** Score buffers and compiled CuTe launches are retained and only rebuilt when a newly admitted request raises the scorer capacity high-water mark.

## Known Limitations

1. PyTorch backend, `KVCacheManagerV2`, and B200/SM100 only.
2. KV-cache block reuse must be disabled.
3. Native sliding-eviction layouts such as Gemma 4 are not supported; the current SWA path targets GPT-OSS-style full-length pools with kernel-side masking.
4. Attention DP and disaggregated serving are not supported.
5. TP beyond TP1 has not yet been validated end to end.

## Test Plan

- Focused TriAttention/KV suite: **55 passed**
  - CuTe score vs. Torch oracle
  - all three eviction modes
  - ragged lengths and unaligned final tokens
  - deterministic heavy-tie settlement
  - target/draft physical co-compaction
  - GPT-OSS-style SWA
  - overlap/page-table staging and cache resize
  - RoPE-fusion gating
  - score scratch larger than 2^31 elements
- Fused-union boundary file: **9/9 passed**
- Focused executor integration: **25 passed**

## Dev Engineer Review

- Added TriAttention KV-cache compression for the PyTorch backend.
- Implemented Triton and SM100 CuTe scoring, selection, tie settlement, and in-place K/V compaction.
- Added configuration, public API, documentation, validation, RoPE-fusion gating, and `IndexMapper.gatherKBlockOffsets` bindings.
- Added target/draft co-compaction for supported speculative modes.
- Scope is limited to PyTorch, `KVCacheManagerV2`, and SM100/B200. Limitations include disabled block reuse, unsupported sliding-eviction layouts, attention DP, disaggregated serving, and TP configurations beyond TP1.

## QA Engineer Review

- Added coverage for offset gathering, manager lifecycle, capability validation, RoPE fusion, CuTe scoring, fused union processing, tie settlement, selection, compaction, page reuse, speculative decoding, and API dispatch.
- Registered `unittest/_torch/kv_cache_compression` in `tests/integration/test_lists/test-db/l0_b200.yml`.
- Reported 55 focused TriAttention/KV tests, 9 fused-union boundary tests, and 25 executor integration tests passing.
- Verdict: **sufficient** for the registered B200 coverage. Unsupported TP, serving, and layout combinations need follow-up coverage if enabled.

### CODE REVIEW

**Review by StanleySun639 — APPROVED** [review]

**Review by tburt-nv — APPROVED** [review]

File: tensorrt_llm/_torch/kv_cache_compression/triattention/triattention.py [hunk 1]
```diff
@@ -0,0 +1,972 @@
...
+    *,
+    request_capacity: int,
+    token_capacity: int,
+) -> Tuple[torch.Tensor, torch.Tensor]:
+    """Allocate the bounded V2 page-table snapshot used by an eviction round."""
+    required_blocks = triton.cdiv(token_capacity, int(manager.tokens_per_block))
+    staged_blocks = min(
+        triton.cdiv(required_blocks, _BLOCK_OFFSET_ALIGNMENT) * _BLOCK_OFFSET_ALIGNMENT,
+        int(manager.max_blocks_per_seq),
+    )
```
[the remark this answers is not in the document]
Comment (liji-nv) [reply tensorrt_llm/_torch/kv_cache_compression/triattention/triattention.py:94]: This concern is valid

**Review by zhaoyangwang-nvidia — APPROVED** [review]
Very interesting technology, approval for runtime part. Need a more detailed review by experts in this area.

File: tensorrt_llm/_torch/kv_cache_compression/triattention/triattention_kernels.py [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+            gathered_rows + (rank * request_count + request) * WIDTH + token,
+            mask=mask,
+            other=-float("inf"),
+        )
+        folded = tl.maximum(folded, value)
+    tl.store(selection_scores_rows + request * WIDTH + token, folded, mask=mask)
+
+
+@triton.jit
+def _settle_ties_kernel(
```
Comment (liji-nv) [inline tensorrt_llm/_torch/kv_cache_compression/triattention/triattention_kernels.py:338]: If this is used outside of the current module, it should not have a `_` prefix. I think usually there is a wrapper function call the triton kernel inside the file.

**Comment by nvpohanh** [comment]
[by Codex] @lowsfer Friendly review reminder: this PR is awaiting your review. Thanks!

File: tensorrt_llm/_torch/kv_cache_compression/triattention/triattention_kernels.py (hunk 2 above)
Comment (Hudayday) [reply tensorrt_llm/_torch/kv_cache_compression/triattention/triattention_kernels.py:338]: Good catch. Fixed these issues.

File: tensorrt_llm/_torch/kv_cache_compression/triattention/triattention.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Hudayday) [reply tensorrt_llm/_torch/kv_cache_compression/triattention/triattention.py:94]: Thanks for flagging this. I double-checked the full V2 path and this case is safe: KVCacheManagerV2 already rounds `max_blocks_per_seq` up to a multiple of 4 for the `PackedInt/int4` copy ABI and allocates the source page table with that width:

  https://github.com/Hudayday/TensorRT-LLM/blob/kvcache-v2-triattention-no-engine/tensorrt_llm/_torch/pyexecutor/kv_cache_manager_v2.py#L1121-L1129

Therefore both operands of `min()` are 4-aligned, so the clamp cannot break alignment. The 4 refers to packed page-table offset entries, not `tokens_per_block`. No change is needed here.

[the remark this answers is not in the document]
Comment (Hudayday) [reply tensorrt_llm/_torch/kv_cache_compression/triattention/triattention.py:94]: @coderabbitai Please verify my comment.

**Review by liji-nv — APPROVED** [review]

File: examples/kv_cache_compression/triattention.md
Comment (lfr-0531) [inline examples/kv_cache_compression/triattention.md:1]: Nit: It'd be better to rename this file to `triattention.md` and change the dir to `kv_cache_compression`.

File: tensorrt_llm/_torch/pyexecutor/_util.py [hunk 3]
```diff
@@ -2262,21 +2262,25 @@ def _create_kv_cache_manager(
...
     """Reject speculative setups the compression method cannot run with."""
-    if (spec_config is None
-            or not config.kv_cache_compression_mode.is_eviction_method()):
+    if spec_config is None:
         return
-    # Evicting methods co-compact the draft KV, so the draft must be a
-    # standard paged cache in the same forward (one-model speculation).
+    if not config.kv_cache_compression_mode.is_eviction_method():
+        return
+    if config.algorithm != "triattention":
```
Comment (lfr-0531) [inline tensorrt_llm/_torch/pyexecutor/_util.py]: Currently, KV cache compression only supports `triattention`, so I don't think it's necessary to add this check here. If we add other methods in the future, we can just add a `support_spec_decoding` flag to `kv_cache_compression_mode`.

File: tensorrt_llm/_torch/pyexecutor/_util.py [hunk 4]
```diff
@@ -2262,21 +2262,25 @@ def _create_kv_cache_manager(
...
     mode = spec_config.spec_dec_mode
     if not (mode.is_mtp_one_model() or mode.is_eagle3_one_model()):
         raise ValueError(
-            f"KV-cache compression algorithm {config.algorithm!r} does not "
-            f"support speculative decoding mode {mode.name}: the draft KV "
-            "must be a standard paged cache compacted together with the "
-            "target (one-model MTP/EAGLE3).")
+            f"KV-cache compression does not support speculative decoding "
+            f"mode {mode.name}; use one-model MTP or EAGLE3")
+    if getattr(config, "eviction_mode", None) != "union":
```
Comment (lfr-0531) [inline tensorrt_llm/_torch/pyexecutor/_util.py]: This should also be included in the `support_spec_decoding` check above.

File: tensorrt_llm/_torch/kv_cache_compression/triattention/triattention.py [hunk 5]
```diff
@@ -0,0 +1,960 @@
...
+import torch
+import triton
+from transformers import AutoConfig
+from transformers.modeling_rope_utils import ROPE_INIT_FUNCTIONS
+
+from ...._utils import prefer_pinned
+from ....bindings.internal.batch_manager.kv_cache_manager_v2_utils import (
+    copy_batch_block_offsets_to_device,
+)
+from ....logger import logger
```
Comment (lfr-0531) [inline tensorrt_llm/_torch/kv_cache_compression/triattention/triattention.py]: We can directly use `from tensorrt_llm.logger import logger` here.

File: tensorrt_llm/_torch/kv_cache_compression/triattention/triattention.py [hunk 6]
```diff
@@ -0,0 +1,960 @@
...
+from .triattention_cute_score_fused import PADDED_HEAD_COLUMNS, build_score_pipeline
+from .triattention_kernels import (
+    fold_union_ranks,
+    gather_mean_phase,
+    reduce_per_head_scores,
+    settle_ties,
+)
+
+if TYPE_CHECKING:
+    from ....llmapi.llm_args import TriAttentionKvCacheCompressionConfig
```
Comment (lfr-0531) [inline tensorrt_llm/_torch/kv_cache_compression/triattention/triattention.py]: Same as above.

**Review by lfr-0531 — APPROVED** [review]
Overall LGTM! Left some minor comments.

File: examples/kv_cache_compression/triattention.md
Comment (Hudayday) [reply examples/kv_cache_compression/triattention.md:1]: Done. Moved the document to examples/kv_cache_compression/triattention.md.

File: tensorrt_llm/_torch/pyexecutor/_util.py (hunk 3 above)
Comment (Hudayday) [reply tensorrt_llm/_torch/pyexecutor/_util.py]: Agree. Removed the TriAttention-specific algorithm check. I also added `supports_speculative_decoding()` as a config capability, so future compression methods can override it without adding algorithm-specific branches to the common validation path.

File: tensorrt_llm/_torch/pyexecutor/_util.py (hunk 4 above)
Comment (Hudayday) [reply tensorrt_llm/_torch/pyexecutor/_util.py]: Done. supports_speculative_decoding() now includes the union-mode requirement.

File: tensorrt_llm/_torch/kv_cache_compression/triattention/triattention.py (hunk 5 above)
Comment (Hudayday) [reply tensorrt_llm/_torch/kv_cache_compression/triattention/triattention.py]: Agree. Updated cross-package imports to use absolute tensorrt_llm imports.

File: tensorrt_llm/_torch/kv_cache_compression/triattention/triattention.py (hunk 6 above)
Comment (Hudayday) [reply tensorrt_llm/_torch/kv_cache_compression/triattention/triattention.py]: Fixed too. Thank you so much!

File: tensorrt_llm/_torch/kv_cache_compression/triattention/triattention.py [hunk 7]
```diff
@@ -0,0 +1,960 @@
...
+                dtype=torch.int32,
+                device=device,
+            )
+
+        self._provisional_rows = torch.zeros(
+            (selection_rows, self.budget), dtype=torch.int32, device=device
+        )
+        self._kept_ordinal_rows = torch.empty_like(self._provisional_rows)
+
+    def _build_score_runtime(
```
Comment (BowenFu) [inline tensorrt_llm/_torch/kv_cache_compression/triattention/triattention.py:855]: Nice work — the manager sits cleanly on KVCacheManagerV2, and keeping compression separate from sparse attention reads well.

One small suggestion: the README lists B200 (SM100) as the validated target and the scoring pipeline goes straight into `sm100_utils`, but I couldn't find a compute-capability check anywhere in the diff. Enabling this on H100 would fail deep in CuTe kernel compilation rather than at config time.

Would an early `get_sm_version()` guard here — raising a plain "TriAttention requires SM100" — be worth adding? It'd turn a confusing build failure into an obvious misconfiguration message, and it matches how the other arch-specific paths gate themselves.

Not blocking on it. Happy to re-review.

**Review by arysef — APPROVED** [review]

**Review by BowenFu — APPROVED** [review]

File: tensorrt_llm/_torch/kv_cache_compression/triattention/triattention.py (hunk 7 above)
Comment (Hudayday) [reply tensorrt_llm/_torch/kv_cache_compression/triattention/triattention.py:855]: Good point. I added an early `is_sm_100f()` guard before the CuTe import, so unsupported GPUs fail clearly while SM100 and SM103 remain supported. Thanks!

**Review by DomBrown — APPROVED** [review]

File: tensorrt_llm/_torch/kv_cache_compression/triattention/triattention.py [hunk 8]
```diff
@@ -0,0 +1,960 @@
...
+            angle = torch.outer(positions + offset, self._omega)
+            cos_table += torch.cos(angle)
+            sin_table += torch.sin(angle)
+        scale = 1.0 / len(_MEAN_PHASE_OFFSETS)
+        self.cos = cos_table.mul_(scale)
+        self.sin = sin_table.mul_(scale)
+        self.rows = target
+
+
+class TriAttention(KVCacheCompressionManager):
```
Comment (yihwang-nv) [inline tensorrt_llm/_torch/kv_cache_compression/triattention/triattention.py]: nit: Can we rename this class? It's looks like an attention backend (eg. TrtllmAttention), but it's a KV cache manager.

`TriAttentionCompressionManager` or `TriAttentionKvCacheManager`, WDYT?

**Review by yihwang-nv — APPROVED** [review]
LGTM! The changes to the attention modules are very straightforward.

File: tensorrt_llm/_torch/kv_cache_compression/triattention/triattention.py (hunk 8 above)
Comment (Hudayday) [reply tensorrt_llm/_torch/kv_cache_compression/triattention/triattention.py]: Good point. I’ll rename it to `TriAttentionCompressionManager`. This keeps the paper/algorithm name while making the runtime role explicit.
